### PR TITLE
MTD validation update

### PR DESCRIPTION
**Pull request by casarsa**
#### PR description:

This is an update of the MTD validation package:

-  the CrossingFrames for the BTL and ETL SimHits have been switched on in order to add the PU SimHits to the SIM monitoring plots;
-  energy and time resolution histograms have been added for BTL;
-  an occupancy plot vs hit energy has been added for BTL.

#### PR validation:

The updated code has been compiled in the release CMSSW_11_0_X_2019-07-24-2300 and tested with the 2026D41 WFs 20434 and 20634 with PU200.

The total size of the MTD monitoring plots is:
476.28 KiB       MTD/ETL
237.66 KiB       MTD/BTL

--------------------------------------------------------------------------------------------------
WF 20434 report (step 3):

MemoryReport> Peak virtual size 4673.77 Mbytes
 Key events increasing vsize: 
[5] run: 1 lumi: 1 event: 5  vsize = 4641.77 deltaVsize = 0.0351562 rss = 3226.27 delta = 29.5234
[2] run: 1 lumi: 1 event: 2  vsize = 4641.74 deltaVsize = 0.304688 rss = 3196.74 delta = 1.77734
[7] run: 1 lumi: 1 event: 7  vsize = 4673.77 deltaVsize = 32 rss = 3273.65 delta = 76.9062
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[9] run: 1 lumi: 1 event: 9  vsize = 4673.77 deltaVsize = 0 rss = 3287.06 delta = 13.4141
[8] run: 1 lumi: 1 event: 8  vsize = 4673.77 deltaVsize = 0 rss = 3287.99 delta = 14.3438
[7] run: 1 lumi: 1 event: 7  vsize = 4673.77 deltaVsize = 32 rss = 3273.65 delta = 47.3828
TimeReport> Time report complete in 506.789 seconds
 Time Summary: 
 - Min event:   1.74964
 - Max event:   12.5869
 - Avg event:   3.23466
 - Total loop:  385.479
 - Total init:  111.365
 - Total job:   506.789
 - EventSetup Lock:   0.000208378
 - EventSetup Get:   84.7993
 Event Throughput: 0.259417 ev/s
 CPU Summary: 
 - Total loop:  367.605
 - Total init:  143.27
 - Total extra: 0
 - Total job:   520.702
 Processing Summary: 
 - Number of Events:  100
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

--------------------------------------------------------------------------------------------------
WF 20634 report (step 3):

MemoryReport> Peak virtual size 9869.82 Mbytes
 Key events increasing vsize: 
[0] run: 0 lumi: 0 event: 0  vsize = 0 deltaVsize = 0 rss = 0 delta = 0
[1] run: 1 lumi: 1 event: 1  vsize = 8169.59 deltaVsize = 0 rss = 6721.41 delta = 0
[2] run: 1 lumi: 1 event: 2  vsize = 9833.69 deltaVsize = 1664.11 rss = 6793.38 delta = 71.9688
[7] run: 1 lumi: 1 event: 7  vsize = 9865.82 deltaVsize = 32 rss = 6828.83 delta = 35.4531
[10] run: 1 lumi: 1 event: 10  vsize = 9869.82 deltaVsize = 4 rss = 6832.15 delta = 3.31641
[8] run: 1 lumi: 1 event: 8  vsize = 9865.82 deltaVsize = 0 rss = 6842.48 delta = 13.6445
[7] run: 1 lumi: 1 event: 7  vsize = 9865.82 deltaVsize = 32 rss = 6828.83 delta = -288.48
[10] run: 1 lumi: 1 event: 10  vsize = 9869.82 deltaVsize = 4 rss = 6832.15 delta = 3.31641
TimeReport> Time report complete in 2705.11 seconds
 Time Summary: 
 - Min event:   137.084
 - Max event:   372.577
 - Avg event:   236.689
 - Total loop:  2488.53
 - Total init:  212.354
 - Total job:   2705.11
 - EventSetup Lock:   0.000253677
 - EventSetup Get:   106.802
 Event Throughput: 0.00401843 ev/s
 CPU Summary: 
 - Total loop:  2186.06
 - Total init:  115.821
 - Total extra: 0
 - Total job:   2306
 Processing Summary: 
 - Number of Events:  10
 - Number of Global Begin Lumi Calls:  1
 - Number of Global Begin Run Calls: 1

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by jfernan2** [comment]
@casarsa I do see plots (6) in BTL but not in ETL like you said in the description. This is happening in the 3 Phase2 workflows. Is this expected?
E.g. https://tinyurl.com/y5po36vr

**Comment by casarsa** [comment]
@jfernan2 For the time being those checks have been implemented only for BTL.

**Comment by jfernan2** [comment]
@casarsa OK, I was confused by your statement: 
"The total size of the MTD monitoring plots is: 
476.28 KiB MTD/ETL "

**Comment by jfernan2** [comment]
+1

**Comment by civanch** [comment]
+1

File: Validation/MtdValidation/plugins/BuildFile.xml [hunk 1]
```diff
@@ -4,4 +4,5 @@
 <use name="Geometry/Records"/>
 <use name="Geometry/MTDGeometryBuilder"/>
 <use name="DQMServices/Core"/>
+<use name="hepmc"/>
```
Comment (fabiocos) [inline Validation/MtdValidation/plugins/BuildFile.xml]: @casarsa why this dependence?

File: Validation/MtdValidation/plugins/BtlRecHitsValidation.cc [hunk 2]
```diff
@@ -96,6 +119,34 @@ void BtlRecHitsValidation::analyze(const edm::Event& iEvent, const edm::EventSet
   const MTDTopology* topology = topologyHandle.product();
 
   auto btlRecHitsHandle = makeValid(iEvent.getHandle(btlRecHitsToken_));
+  auto btlSimHitsHandle = makeValid(iEvent.getHandle(btlSimHitsToken_));
+  MixCollection<PSimHit> btlSimHits(btlSimHitsHandle.product());
+
+  // --- Loop over the BLT SIM hits
+  std::unordered_map<uint32_t, MTDHit> m_btlSimHits;
```
Comment (fabiocos) [inline Validation/MtdValidation/plugins/BtlRecHitsValidation.cc]: @casarsa this energy accumulation is equivalent to that used in the digi - rechit construction, am I correct?

File: Validation/MtdValidation/plugins/BuildFile.xml (hunk 1 above)
Comment (casarsa) [reply Validation/MtdValidation/plugins/BuildFile.xml]: That surprised me too. It seems it's required by SimDataFormats/CrossingFrame/interface/CrossingFrame.h, that includes HepMCProduct.h.

File: Validation/MtdValidation/plugins/BtlRecHitsValidation.cc (hunk 2 above)
Comment (casarsa) [reply Validation/MtdValidation/plugins/BtlRecHitsValidation.cc]: Yes, you are. In both cases I'm starting from the SimHits and am accumulating the energy released in the same BTL crystal.

File: Validation/MtdValidation/plugins/BuildFile.xml (hunk 1 above)
Comment (makortel) [reply Validation/MtdValidation/plugins/BuildFile.xml]: By quick look it looks like the include of `HepMCProduct.h` could easily be removed from `SimDataFormats/CrossingFrame/interface/CrossingFrame.h`.

Comment (makortel) [reply Validation/MtdValidation/plugins/BuildFile.xml]: Regardless of the possible cleanup of `SimDataFormats/CrossingFrame/interface/CrossingFrame.h`, the proper dependence to be added here is `SimDataFormats/CrossingFrame`. That should make the explicit dependence in `hepmc` uneecessary (as `SimDataFormats/CrossingFrame` depends on that).

Comment (kpedro88) [reply Validation/MtdValidation/plugins/BuildFile.xml]: @makortel let's open a separate issue for that

**Comment by kpedro88** [comment]
+1

**Pull request merged**